src/circuit/gadget/ecc/chip/add.rs

```diff
@@ -0,0 +1,482 @@
+        advice_columns.insert(self.beta);
+        advice_columns.insert(self.gamma);
+        advice_columns.insert(self.delta);
+        advice_columns
+    }
+
+    pub(crate) fn create_gate<F: FieldExt>(&self, meta: &mut ConstraintSystem<F>) {
+        let q_add = meta.query_selector(self.q_add, Rotation::cur());
+        let x_p = meta.query_advice(self.x_p, Rotation::cur());
+        let y_p = meta.query_advice(self.y_p, Rotation::cur());
+        let x_q = meta.query_advice(self.x_qr, Rotation::cur());
+        let y_q = meta.query_advice(self.y_qr, Rotation::cur());
+        let x_r = meta.query_advice(self.x_qr, Rotation::next());
+        let y_r = meta.query_advice(self.y_qr, Rotation::next());
+        let lambda = meta.query_advice(self.lambda, Rotation::cur());
+
+        // α = inv0(x_q - x_p)
+        let alpha = meta.query_advice(self.alpha, Rotation::cur());
+        // β = inv0(x_p)
+        let beta = meta.query_advice(self.beta, Rotation::cur());
+        // γ = inv0(x_q)
+        let gamma = meta.query_advice(self.gamma, Rotation::cur());
+        // δ = inv0(y_p + y_q) if x_q = x_p, 0 otherwise
+        let delta = meta.query_advice(self.delta, Rotation::cur());
+
+        // Useful composite expressions
+        // α ⋅(x_q - x_p)
+        let if_alpha = (x_q.clone() - x_p.clone()) * alpha;
+        // β ⋅ x_p
+        let if_beta = x_p.clone() * beta;
+        // γ ⋅ x_q
+        let if_gamma = x_q.clone() * gamma;
+        // δ ⋅(y_p + y_q)
+        let if_delta = (y_q.clone() + y_p.clone()) * delta;
+
+        // Useful constants
+        let one = Expression::Constant(F::one());
+        let two = Expression::Constant(F::from_u64(2));
+        let three = Expression::Constant(F::from_u64(3));
+
+        // Handle cases in incomplete addition
+        {
+            // (x_q − x_p)⋅((x_q − x_p)⋅λ − (y_q−y_p))=0
+            meta.create_gate(
+                "(x_q − x_p)⋅((x_q − x_p)⋅λ − (y_q−y_p))=0",
+                |_| {
+                    let x_q_minus_x_p = x_q.clone() - x_p.clone(); // (x_q − x_p)
+
+                    let y_q_minus_y_p = y_q.clone() - y_p.clone(); // (y_q − y_p)
+                    let incomplete = x_q_minus_x_p.clone() * lambda.clone() - y_q_minus_y_p; // (x_q − x_p)⋅λ − (y_q−y_p)
+
+                    // q_add ⋅(x_q − x_p)⋅((x_q − x_p)⋅λ − (y_q−y_p))
+                    q_add.clone() * x_q_minus_x_p * incomplete
+                },
+            );
+
+            // (1 - (x_q - x_p)⋅α)⋅(2y_p ⋅λ - 3x_p^2) = 0
+            meta.create_gate("(1 - (x_q - x_p)⋅α)⋅(2y_p ⋅λ - 3x_p^2) = 0", |_| {
+                let three_x_p_sq = three * x_p.clone() * x_p.clone(); // 3x_p^2
+                let two_y_p = two.clone() * y_p.clone(); // 2y_p
+                let derivative = two_y_p * lambda.clone() - three_x_p_sq; // (2y_p ⋅λ - 3x_p^2)
+
+                // q_add ⋅(1 - (x_q - x_p)⋅α)⋅(2y_p ⋅λ - 3x_p^2)
+                q_add.clone() * (one.clone() - if_alpha.clone()) * derivative
```

Comment:
2y<sub>p</sub>⋅λ - 3x<sub>p</sub><sup>2</sup> is not the derivative itself; setting it to zero constrains the derivative λ.

```suggestion
                let derivative_equation = two_y_p * lambda.clone() - three_x_p_sq; // (2y_p⋅λ - 3x_p^2)

                // q_add ⋅ (1 - (x_q - x_p)⋅α)⋅(2y_p⋅λ - 3x_p^2)
                q_add.clone() * (one.clone() - if_alpha.clone()) * derivative_equation
```